lib/content-patcher/patchers/class-patcher-abstract.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+namespace NewspackContentConverter\ContentPatcher\Patchers;
+
+use NewspackContentConverter\ContentPatcher\PatcherInterface;
+
+/**
+ * Class PatcherAbstract, containing common functionality.
+ *
+ * @package NewspackContentConverter\ContentPatcher\Patchers
+ */
+abstract class PatcherAbstract implements PatcherInterface {
+
+	/**
+	 * @var array Array of regex expressions -- keys are self explanatory.
+	 *      'match_html_element'
+	 *      'match_blocks_element'
+	 *      'match_attribute_value' - has an extra group, and matches the targeted HTML element attribute's value
+	 *      'replace_blocks_element'
+	 */
+	protected $patterns = [];
+
+	/**
+	 * Runs a preg_match_all with PREG_OFFSET_CAPTURE.
+	 *
+	 * @param string $pattern Pattern.
+	 * @param string $subject Subject.
+	 *
+	 * @return array|null preg_match_all with PREG_OFFSET_CAPTURE matches, or null if no matches.
+	 */
+	protected function match_all_elements( $pattern, $subject ) {
+	    $preg_match_all_result = preg_match_all( $pattern, $subject, $matches, PREG_OFFSET_CAPTURE );
+
+	    return ( false === $preg_match_all_result || 0 === $preg_match_all_result ) ? null : $matches;
+	}
+
+	/**
+	 * Validates the results of match searches.
+	 *
+	 * @param array $matches_html   Regex matches in HTML source.
+	 * @param array $matches_blocks Regex matches in blocks content.
+	 *
+	 * @return bool Do the matches correspond.
+	 */
+	protected function validate_html_and_block_matches( $matches_html, $matches_blocks ) {
+    	if ( count($matches_html) != count($matches_blocks) ) {
+    		return false;
+	    }
+
+    	return true;
+	}
+
+	/**
+	 * Runs a preg_match.
+	 *
+	 * @param string $pattern Pattern.
+	 * @param string $subject Subject.
+	 *
+	 * @return bool|mixed Match result, or false.
+	 */
+	protected function element_match_attribute_value( $pattern, $subject ) {
+		return ( 1 === preg_match( $pattern, $subject, $match ) ) ? $match[ 1 ]: false;
+	}
+
+	/**
+	 * Runs preg_replace.
+	 *
+	 * @param string $pattern_match
+	 * @param string $replacement
+	 * @param string $subject
+	 *
+	 * @return string|string[]|null preg_replace's default return.
+	 */
+	protected function apply_patch_block_element( $pattern_match, $replacement, $subject ) {
```

Comment:
This looks like a pure reimplementation of `preg_replace` - do you foresee implementations of this method that contain more than the one line?

Reply:
Yes, exactly. I think we'll be able to optimize and refactor once we have several more patchers in place, hopefully of different types and scopes, and they'll show us how to improve this. My idea was to keep it simple, point the way, and keep it easily changeable for the next patchers. But flexible and neat, before all.